documentation/specs/multithreaded-msbuild.md

```diff
@@ -0,0 +1,268 @@
+Project2 ->> Scheduler: results
+deactivate Project2
+Scheduler ->> Project1: Project2 outputs
+activate Project1
+Project1 ->> Tasks: Run Task Z
+activate Tasks
+Tasks ->> Project1: Task Z state mutations
+deactivate Tasks
+Project1 ->> Scheduler: results
+deactivate Project1
+```
+
+### Out-of-proc task execution
+
+Even in single-proc mode, MSBuild supports out-of-proc task execution. This is done by spawning a new process to execute the task, which allows the task to run in isolation from the main build process. MSBuild handles passing the necessary information to the task process, such as the current environment, working directory, and explicit inputs. The task process can then execute and return results back to the main build process.
+
+```mermaid
+sequenceDiagram
+
+box Entry Process
+    participant Scheduler
+    participant Project1
+end
+
+Scheduler ->> Project1: start
+activate Project1
+create participant TaskHost
+Project1 -->> TaskHost: Run Task X
+destroy TaskHost
+TaskHost -->> Project1: Task X state mutations
+Project1 ->> Scheduler: results
+deactivate Project1
+```
+
+### Multi-proc builds
+
+In multi-proc builds, the scheduler assigns projects to multiple worker processes (nodes). There is one in-process node running in the entry process, and multiple out-of-process nodes running in separate worker processes, with one node per worker process. Each node is responsible for executing the projects assigned to it. The scheduler manages the creation of the nodes, distribution of projects, and ensures that dependencies are respected. When a project is assigned to a node, it is not moved to another node, which means that the state of the project is maintained within that node.
+
+(This diagram is simplified: the scheduler will not create a new node for every project, so in a two-project build, the scheduler will assign both projects to the same node.)
+
+```mermaid
+sequenceDiagram
+
+box Entry Process
+    participant Scheduler
+    participant Project1
+    participant Tasks
+end
+
+box Worker Process (Node 2)
+    participant Project2
+    participant Tasks2 as Tasks
+end
+
+Scheduler ->> Project1: start
+activate Project1
+Project1 ->> Tasks: Run Task X
+activate Tasks
+Tasks ->> Project1: Task X state mutations
+deactivate Tasks
+Project1 ->> Scheduler: Build Project2
+deactivate Project1
+Scheduler -->> Project2: start
+activate Project2
+Project2 ->> Tasks2: Run Task Y
+activate Tasks2
+Tasks2 ->> Project2: Task Y state mutations
+deactivate Tasks2
+Project2 -->> Scheduler: results
+deactivate Project2
+Scheduler ->> Project1: Project2 outputs
+activate Project1
+Project1 ->> Tasks: Run Task Z
+activate Tasks
+Tasks ->> Project1: Task Z state mutations
+deactivate Tasks
+Project1 ->> Scheduler: results
+deactivate Project1
+```
+
+A task that needs to run out of process can spawn a new process, as in a single-proc build, with potentially many TaskHost processes running concurrently.
+
+
+## Multithreading MSBuild
```

Comment:
why here it is "Multithread**ing** MSBuild" not "Multithread**ed** MSBuild" as the name of the spec?
maybe smth like "Multithreading in MSBuild" for this header?